ledger/src/blockstore.rs

```diff
@@ -301,54 +301,65 @@ impl Blockstore {
         false
     }
 
-    /// Silently deletes all blockstore column families starting at the given slot until the `to` slot
+    /// Silently deletes all blockstore column families in the range [from_slot,to_slot]
     /// Dangerous; Use with care:
     /// Does not check for integrity and does not update slot metas that refer to deleted slots
     /// Modifies multiple column families simultaneously
-    pub fn purge_slots(&self, mut from_slot: Slot, to_slot: Option<Slot>) {
+    pub fn purge_slots_with_delay(
+        &self,
+        from_slot: Slot,
+        to_slot: Slot,
+        delay_between_purges: Option<Duration>,
+    ) {
         // if there's no upper bound, split the purge request into batches of 1000 slots
         const PURGE_BATCH_SIZE: u64 = 1000;
-        let mut batch_end = to_slot.unwrap_or(from_slot + PURGE_BATCH_SIZE);
-        while from_slot < batch_end {
-            match self.run_purge(from_slot, batch_end) {
-                Ok(end) => {
-                    if !self.no_compaction {
-                        if let Err(e) = self.compact_storage(from_slot, batch_end) {
-                            // This error is not fatal and indicates an internal error
-                            error!(
-                                "Error: {:?}; Couldn't compact storage from {:?} to {:?}",
-                                e, from_slot, batch_end
-                            );
-                        }
-                    }
-
-                    if end {
-                        break;
-                    } else {
-                        // update the next batch bounds
-                        from_slot = batch_end;
-                        batch_end = to_slot.unwrap_or(batch_end + PURGE_BATCH_SIZE);
+        let mut batch_start = from_slot;
+        while batch_start < to_slot {
+            let batch_end = (batch_start + PURGE_BATCH_SIZE).min(to_slot);
+            match self.run_purge(batch_start, batch_end) {
+                Ok(_all_columns_purged) => {
+                    batch_start = batch_end;
+
+                    if let Some(ref duration) = delay_between_purges {
+                        // Cooperate with other blockstore users
+                        std::thread::sleep(*duration);
                     }
                 }
                 Err(e) => {
                     error!(
                         "Error: {:?}; Purge failed in range {:?} to {:?}",
-                        e, from_slot, batch_end
+                        e, batch_start, batch_end
                     );
                     break;
                 }
             }
         }
+
+        if !self.no_compaction {
```

Comment:
compact once, for the entire range, instead of on each `run_purge()` iteration.  `run_purge` is fast, `compact_storage` is slow.    The previous code was causing `purge_slots()` to run long enough such that `new_root_receiver` channel fills up with enough roots that the node OOMs